File: 0101-0200/0172.Factorial_Trailing_Zeroes.java
//0ms






class Solution {
    public int trailingZeroes(int n) {
    int res = 0;
    while (n > 0) {
        n /= 5;
        res += n;
    }
    return res;
}}







//1ms






class Solution {
    public int trailingZeroes(int n) {
        int twos = 0, fives = 0;

        // Count powers of 2
        for (int i = 2; Math.pow(2, i) <= n; i++) {
            twos += n / (int)Math.pow(2, i);
        }
        twos += n / 2;

        // Count powers of 5
        for (int i = 2; Math.pow(5, i) <= n; i++) {
            fives += n / (int)Math.pow(5, i);
        }
        fives += n / 5;

        return Math.min(twos, fives);
    }
}

File: 0101-0200/0172.Factorial_Trailing_Zeroes.rs
//0ms






const MAX: usize = 10_001;

const fn trails() -> [usize; MAX] {
    let mut ret = [0; MAX];
    let mut n = 1usize;

    while n < MAX {
        let mut zeroes = 0;
        let mut fives = 5;

        while n / fives > 0 {
            zeroes += n / fives;
            fives *= 5;
        }
        ret[n] = zeroes;
        n += 1;
    }
    ret
}

const TRAILS: [usize; MAX] = trails();

impl Solution {
    pub fn trailing_zeroes(n: i32) -> i32 {
        TRAILS[n as usize] as _
    }
}






//0ms






impl Solution {
  pub fn trailing_zeroes(n: i32) -> i32 {
    if n < 5 {
      return 0;
    }

    let mut count = 0;
    let mut num = n;
    while num >= 5 {
      num /= 5;
      count += num;
    }
    count
  }
}
